kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,382 @@
+	if err != nil {
+		return err
+	}
+
+	// Sort keys.
+	sort.Slice(keys, func(i, j int) bool {
+		return bytes.Compare(keys[i], keys[j]) < 0
+	})
+
+	// Call callback.
+	for _, k := range keys {
+		err = fn(k)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Rollback closes the transaction, discarding changes (if any) if the
+// database was modified by a write transaction.
+func (tx *readWriteTx) Rollback() error {
+	// If the transaction has been closed roolback will fail.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	err := tx.tx.Rollback()
+
+	// Unlock the transaction regardless of the error result.
+	tx.active = false
+	tx.locker.Unlock()
+	return err
+}
+
+// ReadWriteBucket opens the root bucket for read/write access.  If the
+// bucket described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	if len(key) == 0 {
+		return nil
+	}
+
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.NestedReadWriteBucket(key)
+	}
+
+	// Verify that the table containing the specified top-level bucket
+	// exists.
+	var value int
+	err := tx.QueryRow(
+		"SELECT 1 FROM pg_catalog.pg_tables "+
+			"WHERE schemaname='public' AND tablename=$1",
+		table,
+	).Scan(&value)
+
+	switch {
+	case err == sql.ErrNoRows:
+		return nil
+
+	case err != nil:
+		panic(err)
+	}
+
+	return newReadWriteBucket(tx, table, nil)
+}
+
+// getCreateTableSql returns the create statements to set up a kv table in
+// postgres. Every row points to the bucket that it is one via its parent_id
+// field. A NULL parent_id means that the key belongs to the upper-most bucket
+// in this table. A constraint on parent_id is enforcing referential integrity.
+//
+// Furthermore there is a <table>_p index on parent_id that is required for the
+// foreign key constraint.
+//
+// Finally there are unique indices on (parent_id, key) to prevent the same key
+// being present in a bucket more than once (<table>_up and <table>_unp). In
+// postgres, a single index wouldn't enforce the unique constraint on rows with
+// a NULL parent_id. Therefore two indices are defined.
+func getCreateTableSql(table string) string {
+	return `
+CREATE TABLE IF NOT EXISTS public.` + table + `
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT ` + table + `_parent FOREIGN KEY (parent_id)
+        REFERENCES public.` + table + ` (id)
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS ` + table + `_p
+    ON public.` + table + ` (parent_id);
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_up
+    ON public.` + table + `
+    (parent_id, key) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_unp 
```

Comment:
Indexes 🤩

Reply:
Imagine what more indexes we could have when moving from kv to structured sql data...